File: projects/07/VMTranslator/src/com/stockxit/nand2tetris/Parser.java
package com.stockxit.nand2tetris;

import java.util.Scanner;

public class Parser {
    private Scanner mScanner;
    private String mCurrentCommand;

    public Parser(Scanner scanner) {
        this.mScanner = scanner;
    }

    public boolean hasMoreCommands() {
        return mScanner.hasNextLine();
    }

    public void advance() {
        if (hasMoreCommands()) {
            mCurrentCommand = mScanner.nextLine();

            // strip out comment
            int commentIndex = mCurrentCommand.indexOf("/");
            if (commentIndex >= 0) {
                mCurrentCommand = mCurrentCommand.substring(0, commentIndex);
            }

            // trim just to be sure
            mCurrentCommand = mCurrentCommand.trim();

            if (mCurrentCommand.isEmpty()) {
                advance();
            }
        }
    }

    public CommandType commandType() {
        String command = mCurrentCommand.split(" ")[0];
        switch (command) {
            case "push":
                return CommandType.C_PUSH;
            case "pop":
                return CommandType.C_POP;
            default:
                return CommandType.C_ARITHMETIC;
        }
    }

    public String arg1() {
        if (commandType() == CommandType.C_RETURN) {
            return null;
        }

        if (commandType() == CommandType.C_ARITHMETIC) {
            return mCurrentCommand;
        }

        return mCurrentCommand.split(" ")[1];
    }

    public int arg2() {
        if (commandType() == CommandType.C_PUSH
                || commandType() == CommandType.C_POP
                || commandType() == CommandType.C_FUNCTION
                || commandType() == CommandType.C_CALL) {
            return Integer.valueOf(mCurrentCommand.split(" ")[2]);
        }

        return 0;
    }
}

File: projects/06/HackAssembler/src/Parser.kt
import java.util.*

/**
 * Created by thomasstockx on 10/06/2017.
 */
class Parser(var scanner: Scanner) {

    var currentCommand = ""

    var myMap = HashMap<Int,String>()

    fun hasMoreCommands(): Boolean {
        return scanner.hasNextLine()
    }

    fun reset() {
        scanner = scanner.reset()
    }

    fun advance() {
        if (hasMoreCommands()) {
            currentCommand = scanner.nextLine()

            // strip out comment
            var commentIndex = currentCommand.indexOf("/")
            if (commentIndex >= 0) {
                currentCommand = currentCommand.substring(0, commentIndex)
            }

            // strip out whitespace
            currentCommand = currentCommand.replace(" ", "")

            if (currentCommand.isBlank()) {
                advance()
            }
        }
    }

    fun commandType(): CommandType {
        if (currentCommand.startsWith("@")) {
            return CommandType.A_COMMAND
        }
        if (currentCommand.startsWith("(")) {
            return CommandType.L_COMMAND
        }
        return CommandType.C_COMMAND
    }

    fun symbol(): String {
        if(commandType() == CommandType.A_COMMAND) {
            return currentCommand.substring(1)
        }

        if(commandType() == CommandType.L_COMMAND) {
            return currentCommand.substring(1, currentCommand.length - 1)
        }

        return ""
    }

    fun dest(): String {
        if (commandType() == CommandType.C_COMMAND) {
            var temp = currentCommand.split("=")[0]
            return if (currentCommand.contains(";")) "" else temp
        }

        return ""
    }

    fun comp(): String {
        if (commandType() == CommandType.C_COMMAND) {
            var temp = if (currentCommand.contains("=")) currentCommand.split("=")[1] else currentCommand
            return temp.split(";")[0]
        }

        return ""
    }

    fun jump(): String {
        if (commandType() == CommandType.C_COMMAND && currentCommand.contains(";")) {
            return currentCommand.split(";")[1]
        }

        return ""
    }

    fun integerToBinary(integer: String): String {
        var binary = Integer.toBinaryString(Integer.parseInt(integer))
        while (binary.length < 15) {
            binary = "0" + binary
        }
        return binary
    }


}